Fix ValueCheckerUtils.getValuesFromRange

panacekcz opened the pull request:
The method `ValueCheckerUtils.getValuesFromRange` contains code that does not work properly. Furthemore, the method may cause an `OutOfMemoryError` if the upper bound of the input range is `Long.MAX_VALUE`, because in that case the condition of the loop `value<=range.to` always succeeds and the loop variable `value` overflows. Then the list of values grows until memory is exhausted.
This problem can be reproduced by running the value checker on this code:
```java
import org.checkerframework.common.value.qual.*;
public class LongMax {
 public void longMaxRange() {
  @IntRange(from = 9223372036854775807l, to = 9223372036854775807l) long i = 9223372036854775807l;
  @IntRange(from = 9223372036854775807l, to = 9223372036854775807l) long j = i;
 }
}
```
Run the checker by:
```
checker/bin/javac -processor org.checkerframework.common.value.ValueChecker LongMax.java
```
The error output is:
```
error: SourceChecker.typeProcess: unexpected Throwable (OutOfMemoryError) while processing LongMax.java; message: Java heap space; invoke the compiler with -AprintErrorStack to see the stack trace.
```

This PR changes the implementation to:
- not cause `OutOfMemoryError` if `range.to` is `Long.MAX_VALUE`. This is done by looping from 0 to `range.to-range.from` instead of looping from `range.from` to `range.to`. This makes the upper bound of the loop very small (limited by `ValueAnnotatedTypeFactory.MAX_VALUES`), so the loop variable will not overflow before hitting the bound.
- support conversion to `Integer`, `Short`, `Byte` and `Float`
- remove support (which didn't work) for primitive types
- avoid unnecessary boxing and unboxing

An earlier similar change of this method (not fixing the out-of-memory error) was a part of #1352. This PR should be merged before #1352, which requires these fixes.

wmdietlGC commented on framework/src/org/checkerframework/common/value/ValueCheckerUtils.java:
```diff
@@ -134,11 +134,10 @@ public static Range getRangeFromValues(List<? extends Number> values) {
      *
      * @param range the given range
-     * @param expectedType the expected type
+     * @param expectedType the expected type (must be a boxed type, not a primitive type)
```
Can you make this a separate paragraph and explain, why primitive types don't work?
I guess it's because a `List` can only contain boxed types, but it would be good to make explicit.

wmdietlGC commented on framework/src/org/checkerframework/common/value/ValueCheckerUtils.java:
```diff
@@ -149,31 +148,35 @@ public static Range getRangeFromValues(List<? extends Number> values) {
+        // Each value is computed as a sum of the first value and an offset within the range,
+        // to avoid comparing to range.to, which may be Long.MAX_VALUE
+        for (long offset = 0; offset <= length; offset++) {
+            long value = range.from + offset;
```
I don't see how this rewrite is any different than the logic before and this new code has an additional addition.
If `range.from` is `0` and `range.to` is `Long.MAX_VALUE`, both versions iterate over `Long.MAX_VALUE` elements. What is the benefit of using this offset?

panacekcz replied:
This code is executed only for ranges not wider than `ValueAnnotatedTypeFactory.MAX_VALUES`, so here the upper bound of the loop is always a very small number. I will change the comments to explain that. Or I can implement it in a different way if you think that would be better.

wmdietlGC replied:
What is the difference between going from `range.from` to `range.to` and going from `0` to `range.to - range.from` and then adding `range.from` to each value?
In the PR description, can you add what problem you are trying to fix and be explicit about what goes wrong?

wmdietlGC commented on framework/tests/value/LongMax.java:
```diff
@@ -0,0 +1,9 @@
+import org.checkerframework.common.value.qual.*;
+// Test for handling value annotations involving Long.MAX_VALUE
```
Can you describe what fails before this fix?

panacekcz commented:
I have addressed the comments and questions in the code and updated the PR description to explain the problem.

wmdietlGC commented on framework/tests/value/LongMax.java:
```diff
@@ -0,0 +1,15 @@
+import org.checkerframework.common.value.qual.*;
```
Only import the IntRange annotation you actually need.

wmdietlGC commented on framework/src/org/checkerframework/common/value/ValueCheckerUtils.java:
```diff
@@ -149,31 +151,38 @@ public static Range getRangeFromValues(List<? extends Number> values) {
-        } else {
-            throw new UnsupportedOperationException(
-                    "ValueCheckerUtils: unexpected class: " + expectedType);
+            values.add(expectedType.cast(convertedValue));
```
Why do you need to do both the cast in the if-checks above and this `expectedType.cast`?
Can't you keep the basic structure of the previous code, simply removing the primitive cases?
Having the conditions outside of the loop would also make the code more readable. The duplication of the loop logic doesn't seem so bad.

panacekcz replied:
`Class.cast` throws `ClassCastException` if the argument is not of the correct type, so a cast is needed to perform the conversion. `Class.cast` is used for casting to `T` without needing to suppress an unchecked cast warning.
I thought it would be too much duplication to have the loop 7 times there. If this code is not readable enough, I think extracting conversion logic to a separate method should help.

wmdietlGC commented:
Thanks, the PR description is much clearer now!

panacekcz commented:
I have addressed the comments and tried to make the code more readable by extracting the conversion to a separate method. If you still think duplicating the loop would be better, here the code: https://github.com/panacekcz/checker-framework/commit/cbc0c3c1bc7880fa73b5d8f0bc91c5dc901480f1
